video/out/wayland_common.c

```diff
@@ -1297,3 +1314,32 @@ bool vo_wayland_config (struct vo *vo, uint32_t flags)
 
     return true;
 }
+
+void vo_wayland_request_frame(struct vo *vo, void *data, vo_wayland_frame_cb cb)
+{
+    struct vo_wayland_state *wl = vo->wayland;
+    wl->frame.data = data;
+    wl->frame.function = cb;
+    MP_DBG(wl, "restart frame callback\n");
+    frame_callback(wl, NULL, 0);
+}
+
+bool vo_wayland_wait_frame(struct vo *vo)
+{
+    struct vo_wayland_state *wl = vo->wayland;
+
+    if (!wl->frame.callback || wl->frame.dropping)
+        return false;
+
+    // Most likely mpv isn't receiving frame callbacks if the 100ms timeout is hit.
+    // This usually means mpv window is not visible and compositor tells it kindly to not draw anything.
+    int64_t timeout = 100;
+    int64_t start = mp_time_us();
+    while (!wl->frame.pending && vo_wayland_poll(vo, timeout) > 0 && (mp_time_us() - start) / 1000 < timeout);
```

Comment:
I think this should use the time since the last framecallback was received, not the time from _now_.

Also, this should pass the exact waiting time needed to reach the timeout, instead of calling poll with a fixed timeout, and checking whether the timeout was reached afterwards. (Instead, stop looping when the timeout argument to the last poll call was 0.)

Lines longer than 80 columns should be wrapped. Instead of ``while(something);`` I think it's better to do at least ``while(something) {}`` to make it really clear that the body is empty.